src/dataflow/operators/map_operator.rs

```diff
@@ -1,99 +1,78 @@
-/// let u64_stream = connect_1_write!(MapOperator<u32, u64>, map_config, u32_stream);
+/// fn square_fn(a: &usize) -> usize {
+///     return a * a;
+/// }
+/// # Add the mapping function as an argument to the operator via the OperatorConfig.
+/// let map_config = OperatorConfig::new().name("MapOperator");
+/// let map_stream = erdos::connect_one_in_one_out(
+///     || -> MapOperator<usize, usize> { MapOperator::new(square_fn) },
+///     || {},
+///     map_config,
+///     &source_stream,
+/// );
 /// ```
-pub struct MapOperator<D1: Data, D2: Data> {
-    phantom_data: PhantomData<(D1, D2)>,
+#[derive(Clone)]
+pub struct MapOperator<D1, D2>
+where
+    D1: Data + for<'a> Deserialize<'a>,
+    D2: Data + for<'a> Deserialize<'a>,
+{
+    map_function: Arc<dyn MapFn<D1, D2>>,
 }
 
-impl<'a, D1: Data, D2: Data + Deserialize<'a>> MapOperator<D1, D2> {
-    /// Returns a new instance of the MapOperator.
-    ///
-    /// # Arguments
-    /// * `config` - An instance of OperatorConfig that provides the closure used to map items of
-    /// type D1 to D2.
-    /// * `input_stream` - Represents the incoming stream of messages of type D1.
-    /// * `output_stream` - Represents an outgoing stream of messages of type D2.
-    pub fn new<F: 'static + Clone + Fn(&D1) -> D2>(
-        config: OperatorConfig<F>,
-        input_stream: ReadStream<D1>,
-        output_stream: WriteStream<D2>,
-    ) -> Self {
-        // TODO :: We do this because otherwise we would either have to clone the output stream or
-        // mutex the original output stream. This code should be fixed once add_callback passes
-        // output streams by default.
-        let stateful_stream = input_stream.add_state(output_stream);
-
-        // Clone the name so that we can move the passed function into the callback.
-        let name: String = config
-            .name
-            .clone()
-            .unwrap_or_else(|| format!("MapOperator {}", config.id));
-        let callback = config
-            .arg
-            .unwrap_or_else(|| panic!("{}: no map function supplied", name));
-
-        stateful_stream.add_callback(
-            move |t: &Timestamp, msg: &D1, output_stream: &mut WriteStream<_>| {
-                Self::on_data_callback(t, msg, output_stream, &callback)
-            },
-        );
+impl<D1, D2> MapOperator<D1, D2>
+where
+    D1: Data + for<'a> Deserialize<'a>,
+    D2: Data + for<'a> Deserialize<'a>,
+{
+    pub fn new(map_function: impl MapFn<D1, D2> + 'static) -> Self {
         Self {
-            phantom_data: PhantomData,
+            map_function: Arc::new(map_function),
         }
     }
+}
 
-    /// Returns a new instance of a WriteStream to send its outgoing messages on.
-    ///
-    /// # Arguments
-    /// * `input_stream` - Represents the incoming stream of messages of type D1.
-    pub fn connect(_input_stream: &ReadStream<D1>) -> WriteStream<D2> {
-        WriteStream::new()
+impl<D1, D2> OneInOneOut<(), D1, D2> for MapOperator<D1, D2>
```

Comment:
Can you please implement the `ParallelOneInOneOut` interface here? That would require you to also implement `AppendableStateT` for `()` in `src/dataflow/state.rs`.